bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/AbstractRegistry.java

```diff
@@ -268,6 +279,15 @@ protected void addProvider(Provider<E> provider) {
         }
     }
 
+    /**
+     * Returns the key for a given element
+     *
+     * @param element
+     *            element
+     * @return key (must not be null)
+     */
+    protected abstract K getKey(E element);
```

Comment:
If we parameterize the Provider, then we'll need to introduce a new method "K getKey(E element)" in the Provider interface and then the registries should ask the providers about the key for the element and the control over the key will become decentralized. Do we need such freedom? Also this will be larger change against this proposal.